docs/docs/how-to/local-development/graphql-typegen.md

```diff
@@ -0,0 +1,224 @@
+---
+title: GraphQL Typegen
+examples:
+  - label: Using GraphQL Typegen
+    href: "https://github.com/gatsbyjs/gatsby/tree/master/examples/using-graphql-typegen"
+---
+
+## Introduction
+
+If you're already using [Gatsby with TypeScript](/docs/how-to/custom-configuration/typescript) and manually typing the results of your queries, you'll learn in this guide how Gatsby's automatic GraphQL Typegen feature can make your life easier. By relying on the types that are generated by Gatsby itself and using autocompletion for GraphQL queries in your IDE you'll be able to write GraphQL queries quicker and safer.
+
+This feature was added in `gatsby@4.15.0`.
+
+## Prerequisites
+
+- A Gatsby project set up with `gatsby@4.15.0` or later.
+- The config option [`graphqlTypegen`](/docs/reference/config-files/gatsby-config/#graphqltypegen) set to `true` inside `gatsby-config`.
+  ```js:title=gatsby-config.js
+  module.exports = {
+    graphqlTypegen: true,
+  }
+  ```
+- A `tsconfig.json` in your project with `"include": ["./src/**/*"]` set. See [full `tsconfig.json` example](https://github.com/gatsbyjs/gatsby/tree/master/examples/using-graphql-typegen/tsconfig.json).
+- Optional: If you use VSCode you may install the [GraphQL extension](https://marketplace.visualstudio.com/items?itemName=GraphQL.vscode-graphql).
+
+## Using the autogenerated `Queries` types
+
+For this example to work you'll have to have a `title` inside your `siteMetadata` (in `gatsby-config`).
+
+1. Start the development server with `gatsby develop`. Once the server is ready you should see a log message `Generating GraphQL and TypeScript types` at the bottom of your terminal.
+1. Gatsby created a type generation file which you should see at `src/gatsby-types.d.ts`. It contains the TypeScript types for your queries. Your `tsconfig.json` should include this file so that you can access the `namespace Queries` everywhere in your project.
+1. Create a new page at `src/pages/typegen.tsx` with the following contents:
+
+   ```tsx
+   import * as React from "react"
+   import { graphql, PageProps } from "gatsby"
+
+   const TypegenPage = ({ data }: PageProps) => {
+     return (
+       <main style={pageStyles}>
+         <p>Site title: TODO</p>
+         <hr />
+         <p>Query Result:</p>
+         <pre>
+           <code>{JSON.stringify(data, null, 2)}</code>
+         </pre>
+       </main>
+     )
+   }
+
+   export default TypegenPage
+
+   export const query = graphql`
+     query TypegenPage {
+       site {
+         siteMetadata {
+           title
+         }
+       }
+     }
+   `
+   ```
+
+   It is important that your query has a name (here: `query TypegenPage {}`) as otherwise the automatic type generation doesn't work. We recommend naming the query the same as your React component and using [PascalCase](https://en.wiktionary.org/wiki/Pascal_case). You can enforce this requirement by using [`graphql-eslint`](#graphql-eslint).
```

Comment:
extra space to remove here

Reply:
I intentionally indented it here as it refers to the code block above